README
NAME
    Catalyst::View::ContentNegotiation::XHTML - Adjusts the response
    Content-Type header to application/xhtml+xml if the browser accepts it.

SYNOPSIS
        package Catalyst::View::TT;

        use Moose;
        use namespace::clean -except => 'meta';

        extends qw/Catalyst::View::TT/;
        with qw/Catalyst::View::ContentNegotiation::XHTML/;

        1;

DESCRIPTION
    This is a simple Role which sets the response "Content-Type" to be
    "application/xhtml+xml" if the users browser sends an "Accept" header
    indicating that it is willing to process that MIME type.

    Changing the "Content-Type" to "application/xhtml+xml" causes browsers
    to interpret the page as XML, meaning that your markup must be well
    formed.

CAVEATS
    This is useful when you're developing your application, as you know that
    all pages you view are parsed as XML, so any errors caused by your
    markup not being well-formed will show up at once.

    Whilst this module is has been tested against most popular browsers
    including Internet Explorer, it may cause unexpected results on browsers
    which do not properly support the "application/xhtml+xml" MIME type.

METHOD MODIFIERS
  after process
    Changes the response "Content-Type" if appropriate (from the requests
    "Accept" header).

METHODS
  pragmatic_accept
    Some browsers (such as Internet Explorer) have a nasty way of sending
    Accept */* and this claiming to support XHTML just as well as HTML.
    Saving to a file on disk or opening with another application does count
    as accepting, but it really should have a lower q value then text/html.
    This sub takes a pragmatic approach and corrects this mistake by
    modifying the Accept header before passing it to content negotiation.

ATTRIBUTES
  variants
    Returns an array ref of 3 part arrays, comprising name, priority, output
    mime-type, which is used for the content negotiation algorithm.

PRIVATE METHODS
  _build_variants
    Returns the default variant attribute contents.

SEE ALSO
    Catalyst::View::TT::XHTML - Trivial Catalyst TT view using this role.
    <http://www.w3.org/Protocols/rfc2616/rfc2616-sec12.html> - Content
    negotiation RFC.

BUGS
    Should be split into a base ContentNegotiation role which is consumed by
    ContentNegotiation::XHTML.

AUTHOR
    Original author and maintainer - Tomas Doran (t0m)
    "<bobtfish@bobtfish.net>"

CONTRIBUTORS
    David Dorward - test patches and */* pragmatism to make it work for
    browsers which aren't firefox.
    Florian Ragwitz (rafl) "<rafl@debian.org>" - Conversion into a Moose
    Role, which is what the module should have been originally.

COPYRIGHT
    This module itself is copyright (c) 2008 Tomas Doran and is licensed
    under the same terms as Perl itself.



Changes
1.102
  - Add Vary header to prevent potential problems with downstream caches
1.101
  - Fix dependencies in Makefile.PL, merge fail :/
1.100
  - Refactor into a Moose Role for use with alternate views. (rafl)
  - Additional documentation (t0m)
  - Clone request headers rather than changing them (t0m)
1.004
  - Nick the OSX fragment out of the Catalyst::Runtime Makefile.PL to
    beat my Mac into generating a correct dist. (t0m)
1.003
  - Fixes an tests to be fully Internet Explorer compatible (David Dorward)
  - Change to MRO::Compat for perl 5.10 (t0m)
1.002 2008-12-13
  - Add 'use Class::C3' so that the module works on the currently 
    released Catalyst version. (t0m)
1.001 2008-12-12
  - Add tests for other Accept header cases where the current code
    will get it wrong. (David Dorward)
  - Fix all of these tests. (t0m)
1.000 2008-12-12
  - First working version of the module extracted from the quick hack
    I have in every Catalyst application I've ever written. (t0m)


t/lib/TestApp/View/XHTML.pm
package TestApp::View::XHTML;
use strict;
use warnings;
use base qw/Catalyst::View::TT::XHTML/;

1;


t/lib/TestApp/Controller/Root.pm
package TestApp::Controller::Root;
use strict;
use warnings;

__PACKAGE__->config(namespace => q{});

use base 'Catalyst::Controller';

sub main :Path { }

sub nothtml :Local {
  my ($self, $c) = @_;
  $c->res->content_type('application/json');
}

sub end : ActionClass('RenderView') {}

1;


Makefile.PL
use inc::Module::Install 0.91;
use strict;
use warnings;

BEGIN {
    if ($Module::Install::AUTHOR) {
        require Module::Install::AuthorRequires;
        require Module::Install::AuthorTests;
    }
}

name 'Catalyst-View-ContentNegotiation-XHTML';
all_from 'lib/Catalyst/View/ContentNegotiation/XHTML.pm';

requires 'Catalyst::Runtime';
requires 'Catalyst::View::TT';
requires 'HTTP::Negotiate';
requires 'MRO::Compat';
requires 'MooseX::Types::Structured';
requires 'MooseX::Types::Moose';

test_requires 'Catalyst::Action::RenderView';
test_requires 'Test::WWW::Mechanize::Catalyst';
test_requires 'Test::More';

if ($Module::Install::AUTHOR) {
  author_requires 'Test::Pod' => 1.14;
  author_requires 'Test::Pod::Coverage' => 1.04;

  if ($^O eq 'darwin') {
      my $osx_ver = `/usr/bin/sw_vers -productVersion`;
      chomp $osx_ver;

      # TAR on 10.4 wants COPY_EXTENDED_ATTRIBUTES_DISABLE
      # On 10.5 (Leopard) it wants COPYFILE_DISABLE
      my $attr = $osx_ver eq '10.5' ? 'COPYFILE_DISABLE' : 'COPY_EXTENDED_ATTRIBUTES_DISABLE';

      makemaker_args(dist => { PREOP => qq{\@if [ "\$\$$attr" != "true" ]; then}.
                                        qq{ echo "You must set the ENV variable $attr to true,"; }.
                                        ' echo "to avoid getting resource forks in your dist."; exit 255; fi' });
  }
  author_tests 't/author';
}
auto_install;
resources repository => 'git://git.shadowcat.co.uk/catagits/Catalyst-View-ContentNegotiation-XHTML.git';

WriteAll();



lib/Catalyst/View/ContentNegotiation/XHTML.pm
package Catalyst::View::ContentNegotiation::XHTML;

use Moose::Role;
use MooseX::Types::Moose qw/Num Str ArrayRef/;
use MooseX::Types::Structured qw/Tuple/;
use HTTP::Negotiate qw/choose/;

use namespace::clean -except => 'meta';

# Remember to bump $VERSION in View::TT::XHTML also.
our $VERSION = '1.102';

requires 'process';

has variants => (
    is      => 'ro',
    isa     => ArrayRef[Tuple[Str, Num, Str]],
    lazy    => 1,
    builder => '_build_variants',
);

sub _build_variants {
    return [
        [qw| xhtml 1.000 application/xhtml+xml |],
        [qw| html  0.900 text/html             |],
    ];
}

after process => sub {
    my ($self, $c) = @_;
    if ( my $accept = $self->pragmatic_accept($c) and $c->response->headers->{'content-type'} =~ m|text/html|) {
        my $headers = $c->request->headers->clone;
        $headers->header('Accept' => $accept);
        if ( choose($self->variants, $headers) eq 'xhtml') {
            $c->response->headers->{'content-type'} =~ s|text/html|application/xhtml+xml|;
        }
    }
    $c->response->headers->push_header(Vary => 'Accept');
};

sub pragmatic_accept {
    my ($self, $c) = @_;
    my $accept = $c->request->header('Accept') or return;
    if ($accept =~ m|text/html|) {
        $accept =~ s!\*/\*\s*([,]+|$)!*/*;q=0.5$1!;
    } 
    else {
        $accept =~ s!\*/\*\s*([,]+|$)!text/html,*/*;q=0.5$1!;
    }
    return $accept;
}

1;

__END__

=head1 NAME

Catalyst::View::ContentNegotiation::XHTML - Adjusts the response Content-Type
header to application/xhtml+xml if the browser accepts it.

=head1 SYNOPSIS

    package Catalyst::View::TT;

    use Moose;
    use namespace::clean -except => 'meta';

    extends qw/Catalyst::View::TT/;
    with qw/Catalyst::View::ContentNegotiation::XHTML/;

    1;

=head1 DESCRIPTION

This is a simple Role which sets the response C<Content-Type> to be
C<application/xhtml+xml> if the users browser sends an C<Accept> header
indicating that it is willing to process that MIME type.

Changing the C<Content-Type> to C<application/xhtml+xml> causes browsers to
interpret the page as XML, meaning that your markup must be well formed.

=head1 CAVEATS

This is useful when you're developing your application, as you know that all
pages you view are parsed as XML, so any errors caused by your markup not
being well-formed will show up at once.

Whilst this module is has been tested against most popular browsers including
Internet Explorer, it may cause unexpected results on browsers which do not
properly support the C<application/xhtml+xml> MIME type.

=head1 METHOD MODIFIERS

=head2 after process

Changes the response C<Content-Type> if appropriate (from the requests
C<Accept> header).

=head1 METHODS

=head2 pragmatic_accept

Some browsers (such as Internet Explorer) have a nasty way of sending Accept
*/* and this claiming to support XHTML just as well as HTML. Saving to a file
on disk or opening with another application does count as accepting, but it
really should have a lower q value then text/html. This sub takes a pragmatic
approach and corrects this mistake by modifying the Accept header before
passing it to content negotiation.

=head1 ATTRIBUTES

=head2 variants

Returns an array ref of 3 part arrays, comprising name, priority, output
mime-type, which is used for the content negotiation algorithm.

=head1 PRIVATE METHODS

=head2 _build_variants

Returns the default variant attribute contents.

=head1 SEE ALSO

=over

=item L<Catalyst::View::TT::XHTML> - Trivial Catalyst TT view using this role.

=item L<http://www.w3.org/Protocols/rfc2616/rfc2616-sec12.html> - Content
negotiation RFC.

=back

=head1 BUGS

Should be split into a base ContentNegotiation role which is consumed by
ContentNegotiation::XHTML.

=head1 AUTHOR

    Maintainer and contributor of various features - David Dorward (dorward) C<< <david@dorward.me.uk> >>

    Original author and maintainer - Tomas Doran (t0m) C<< <bobtfish@bobtfish.net> >>

=head1 CONTRIBUTORS

=over

=item Florian Ragwitz (rafl) C<< <rafl@debian.org> >> - Conversion into a
Moose Role, which is what the module should have been originally.

=back

=head1 COPYRIGHT

This module itself is copyright (c) 2008 Tomas Doran and is licensed under the
same terms as Perl itself.

=cut
